pkg/solana/client/client.go

```diff
@@ -0,0 +1,152 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/config"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/logger"
+	"golang.org/x/sync/singleflight"
+)
+
+//go:generate mockery --name ReaderWriter --output ./mocks/
+type ReaderWriter interface {
+	Writer
+	Reader
+}
+
+type Reader interface {
+	AccountReader
+	Balance(addr solana.PublicKey) (uint64, error)
+	SlotHeight() (uint64, error)
+	LatestBlockhash() (*rpc.GetLatestBlockhashResult, error)
+	ChainID() (string, error)
+	GetFeeForMessage(msg string) (uint64, error)
+}
+
+// AccountReader is an interface that allows users to pass either the solana rpc client or the relay client
+type AccountReader interface {
+	GetAccountInfoWithOpts(ctx context.Context, addr solana.PublicKey, opts *rpc.GetAccountInfoOpts) (*rpc.GetAccountInfoResult, error)
+}
+
+type Writer interface {
+	SendTx(ctx context.Context, tx *solana.Transaction) (solana.Signature, error)
+}
+
+var _ ReaderWriter = (*Client)(nil)
+
+type Client struct {
+	rpc             *rpc.Client
+	skipPreflight   bool // to enable or disable preflight checks
+	commitment      rpc.CommitmentType
+	txTimeout       time.Duration
+	contextDuration time.Duration
+	log             logger.Logger
+
+	// provides a duplicate function call suppression mechanism
```

Comment:
is this still needed? Seems like we already have caching at a layer higher and this just makes things harder to reason about

Reply:
i think it still is, the caching is only for the state pollling but doesn't include anything like blockhash, balances, etc

Reply:
are there rpc problems if we do lots of blockhash/balance queries though?

Reply:
we have yet to hit a limit, but i suspect that it could become an issue as we try to scale the number of feeds and decrease the round time.

right now we're querying state (3 calls per second per feed) + transmitting (2 calls blockhash + transmit, per feed) + querying balance (1 call per 5 second, separate)
and in the future we'd probably need to go faster on the state and transmitting calls...so having coalescing on certain calls seems to be useful?

Reply:
the coalescing feels like a crutch which masks the real load were sending and the load the node can handle. Imo we should confirm the exact query load the node can handle, then update our cache refresh period and add explicit rate limiting accordingly

Reply:
Connor, do you see any potential issues if the client actually coalesces requests?

I do agree that most of the caching, as currently implemented, is on the higher level, and we'll probably coalesce only `blockhash` calls while running feeds. But I don't see any problems if the lower-level client actually contributes to the overall performance optimization of the system in this way, do you?

Not to mention other parts of the system like monitoring, e2e tests, which should ideally all use a single (type, not instance) functional and robust client. Those systems can then execute parallel calls to the client, without worrying about call de/duplication.

In any case, as we have larger performance efforts incoming, I'd suggest we leave it as it was for now and address it if necessary in the follow-up (performance-focused PR).

Reply:
I don't think there are any implications for correctness with the overall system right now, so fine to leave as is. My only concern with it is just unjustified complexity, I think having a dumb client and caching a layer above is sufficient for our use cases. I think it'd be totally fine to add it to the client if the requests it was handling were stateless (which is I think its main use case), but they are not, so it runs the risk of returning staler than necessary blockhashes/nonces etc.